plugins/cros-ec/fu-cros-ec-usb-device.c

```diff
@@ -0,0 +1,327 @@
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (device);
+	FuCrosEcUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+
+	if (!g_usb_device_claim_interface (usb_device, self->iface_idx,
+					   G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
+					   error)) {
+		g_prefix_error (error, "failed to claim interface: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+
+static gboolean
+fu_cros_ec_usb_device_probe (FuUsbDevice *device, GError **error)
+{
+	FuCrosEcUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+
+	/* very much like usb_updater2's usb_findit() */
+
+	if (!fu_cros_ec_usb_device_find_interface (device, error)) {
+		g_prefix_error (error, "failed to find update interface: ");
+		return FALSE;
+	}
+
+	if (self->chunk_len == 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "wMaxPacketSize isn't valid: %" G_GUINT16_FORMAT,
+			     self->chunk_len);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+do_xfer (FuCrosEcUsbDevice * cros_ec_device, guint8 *outbuf, gsize outlen,
+	 guint8 *inbuf, gsize inlen, gboolean allow_less,
+	 gsize *rxed_count, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (cros_ec_device));
+	FuCrosEcUsbDevice *self = cros_ec_device;
+	gsize actual = 0;
+
+	/* send data out */
+	if (outbuf != NULL && outlen > 0) {
+		if (!g_usb_device_bulk_transfer (usb_device, self->ep_num,
+						 outbuf, outlen,
+						 &actual, BULK_SEND_TIMEOUT_MS,
+						 NULL, error)) {
+			return FALSE;
+		}
+		if (actual != outlen) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_PARTIAL_INPUT,
+				     "only sent %" G_GSIZE_FORMAT "/%"
+				     G_GSIZE_FORMAT " bytes",
+				     actual, outlen);
+			return FALSE;
+		}
+	}
+
+	/* read reply back */
+	if (inbuf != NULL && inlen > 0) {
+		actual = 0;
+		if (!g_usb_device_bulk_transfer (usb_device,
+						 self->ep_num | 0x80,
+						 inbuf, inlen,
+						 &actual, BULK_RECV_TIMEOUT_MS,
+						 NULL, error)) {
+			return FALSE;
+		}
+		if (actual != inlen && !allow_less) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_PARTIAL_INPUT,
+				     "only received %" G_GSIZE_FORMAT "/%"
+				     G_GSIZE_FORMAT " bytes",
+				     actual, outlen);
+			return FALSE;
+		}
+	}
+
+	if (rxed_count != NULL)
+		*rxed_count = actual;
+
+	return TRUE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_setup (FuDevice *device, GError **error)
+{
+	FuCrosEcUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+	gsize rxed_size;
+	guint32 error_code;
+	struct update_frame_header ufh;
+
+	/*
+	 * Need to be backwards compatible, communicate with targets running
+	 * different protocol versions.
+	 */
+	union {
+		struct first_response_pdu rpdu;
+		guint32 legacy_resp;
+	} start_resp;
+
+	/* send start request */
+
+	/* Flush all data from endpoint to recover in case of error. */
+	/* FIXME(bleung): b/157109635. This flush causes problems with the next
```

Comment:
outside of chrome OS development I don't think b/#### is very well understood.  Assuming this is a public bug number, could you expand this to something resolvable?

That being said having no idea on the underlying architecture, is this just a case that the flush command returns before the content has _actually_ propagated to the other parts in the black box?  If so, a reasonable thing to do would be to throw a loop here that will try some reasonable number of times until the transfers stop timing out.

I do something similar in the Dell dock setup routine where there is a corner race condition that you could potentially try to probe some composite devices before they've finished booting if you plug everything in at once.
https://github.com/fwupd/fwupd/blob/master/plugins/dell-dock/fu-dell-dock-i2c-ec.c#L916

This is also the reason we have the helper function `fu_device_retry` for sending some commands.

Reply:
Hi @superm1,

I can change this to an issue in this public project. I got some comments from the initial developer of usb_updater2.c that I can reflect out to the public issue.

I suspect that there is some underlying difference between my fwupd plugin implementation's use of bulk transfer (which is async) versus, the one used in usb_updater2, which calls the synchronous libusb bulk transfer. We can discuss it in the issue further.

Reply:
https://github.com/fwupd/fwupd/issues/2138 - This was fixed. Thanks!